src/libraries/System.Private.CoreLib/src/System/Text/RunePosition.cs

```diff
@@ -0,0 +1,291 @@
+/// </remarks>
+public readonly struct RunePosition : IEquatable<RunePosition>
+{
+    /// <summary>
+    /// Returns an enumeration of <see cref="RunePosition"/> from the provided span that allows deeper data inspection.
+    /// </summary>
+    /// <param name="span">The <see cref="ReadOnlySpan{T}"/> with Unicode data.</param>
+    /// <returns>
+    /// <see cref="Utf16Enumerator"/> to enumerate <see cref="RunePosition"/> from the provided span with UTF-16
+    /// Unicode data.
+    /// </returns>
+    /// <remarks>
+    /// Invalid Unicode symbols will be represented by <see cref="System.Text.Rune.ReplacementChar"/>
+    /// value.
+    /// </remarks>
+    public static Utf16Enumerator EnumerateUtf16(ReadOnlySpan<char> span) => new(span);
+
+    /// <summary>
+    /// Returns an enumeration of <see cref="RunePosition"/> from the provided span that allows deeper data inspection.
+    /// </summary>
+    /// <param name="span">The <see cref="ReadOnlySpan{T}"/> with Unicode data.</param>
+    /// <returns>
+    /// <see cref="Utf8Enumerator"/> to enumerate <see cref="RunePosition"/> from the provided span with UTF-8 Unicode
+    /// data.
+    /// </returns>
+    /// <remarks>
+    /// Invalid Unicode symbols will be represented by <see cref="Rune.ReplacementChar"/> value.
+    /// </remarks>
+    public static Utf8Enumerator EnumerateUtf8(ReadOnlySpan<byte> span) => new(span);
+
+    /// <summary>
+    /// Unicode scalar value <see cref="System.Text.Rune"/> of the current symbol in Unicode data.
+    /// Invalid Unicode symbols will be represented by <see cref="System.Text.Rune.ReplacementChar"/> value.
+    /// </summary>
+    public Rune Rune { get; }
+
+    /// <summary>
+    /// The index of current symbol in Unicode data.
+    /// </summary>
+    public int StartIndex { get; }
+
+    /// <summary>
+    /// The length of current symbol in Unicode data.
+    /// </summary>
+    public int Length { get; }
+
+    /// <summary>
+    /// <see langword="false"/> it current Unicode symbol is correct encoded and <see cref="RunePosition.Rune"/>
+    /// contain its scalar value.
+    /// <br />
+    /// <see langword="true"/> if current Unicode symbol is invalid encoded and <see cref="RunePosition.Rune"/> was
+    /// replaced by <see cref="System.Text.Rune.ReplacementChar"/> value.
+    /// </summary>
+    public bool WasReplaced { get; }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="RunePosition"/> struct.
+    /// </summary>
+    /// <param name="rune">The Unicode scalar value.</param>
+    /// <param name="startIndex">The index of the current symbol in Unicode data.</param>
+    /// <param name="length">The length of the current symbol in Unicode data.</param>
+    /// <param name="wasReplaced">Indicates if the current Unicode symbol was replaced.</param>
+    internal RunePosition(Rune rune, int startIndex, int length, bool wasReplaced)
+    {
```

Comment:
Looks this what we  need to expose according to the review comment `We should have a constructor that sets all four fields`?